ksqldb-rest-app/src/main/java/io/confluent/ksql/api/impl/BlockingPrintPublisher.java

```diff
@@ -0,0 +1,206 @@
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.api.impl;
+
+import io.confluent.ksql.api.server.KsqlApiException;
+import io.confluent.ksql.parser.tree.PrintTopic;
+import io.confluent.ksql.reactive.BasePublisher;
+import io.confluent.ksql.rest.Errors;
+import io.confluent.ksql.rest.server.resources.streaming.PrintTopicUtil;
+import io.confluent.ksql.rest.server.resources.streaming.RecordFormatter;
+import io.confluent.ksql.services.ServiceContext;
+import io.confluent.ksql.util.KsqlException;
+import io.confluent.ksql.util.VertxUtils;
+import io.vertx.core.Context;
+import io.vertx.core.Future;
+import io.vertx.core.WorkerExecutor;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ExecutionException;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.consumer.ConsumerRecords;
+import org.apache.kafka.clients.consumer.KafkaConsumer;
+import org.apache.kafka.common.record.RecordBatch;
+import org.apache.kafka.common.utils.Bytes;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A query publisher that uses an internal blocking queue to store rows for delivery. It's currently
+ * necessary to use a blocking queue as Kafka Streams delivers message in a synchronous fashion with
+ * no back pressure. If the queue was not blocking then if the subscriber was slow the messages
+ * could build up on the queue eventually resulting in out of memory. The only mechanism we have to
+ * slow streams down is to block the thread. Kafka Streams uses dedicated streams per topology so
+ * this won't prevent the thread from doing useful work elsewhere but it does mean we can't have too
+ * many push queries in the server at any one time as we can end up with a lot of threads.
+ */
+public class BlockingPrintPublisher extends BasePublisher<String> {
+
+  private static final Logger log = LoggerFactory.getLogger(BlockingPrintPublisher.class);
+
+  public static final int SEND_MAX_BATCH_SIZE = 200;
+
+  private final WorkerExecutor workerExecutor;
+  private final KafkaConsumer<Bytes, Bytes> topicConsumer;
+  private final PrintTopic printTopic;
+  private final int limit;
+  private final long interval;
+  private final RecordFormatter formatter;
+  private final Duration disconnectCheckInterval;
+  private volatile boolean closed;
+  private volatile boolean started;
+
+  public BlockingPrintPublisher(final Context ctx,
+      final WorkerExecutor workerExecutor,
+      final ServiceContext serviceContext,
+      final Map<String, Object> consumerProperties,
+      final PrintTopic printTopic,
+      final Duration disconnectCheckInterval) {
```

Comment:
This disconnect check interval is currently coming from KsqlRestConfig#STREAMED_QUERY_DISCONNECT_CHECK_MS_CONFIG, which is used for a very different purpose. Let's not conflate the two. 

Its only usage in this class right now is the topic consumer's poll interval. We can choose a timeout ourselves instead.

Reply:
That sounds reasonable. Considering the interval name, it doesn't fit what we want here, so I'm happy to change it. Should we use a fixed value, something like 50 ms?

A bit of context on why I replicated it like this: basically that's what the current web sockets implementation use as a timeout. 
Here's where the polling happens: 
https://github.com/confluentinc/ksql/blob/8d7f7080dc9742bb1e4bc19904228f2e763e646c/ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/TopicStreamWriter.java#L111

Tracking down the origin, you come to this line, where  STREAMED_QUERY_DISCONNECT_CHECK_MS_CONFIG is used.
https://github.com/confluentinc/ksql/blob/8d7f7080dc9742bb1e4bc19904228f2e763e646c/ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/KsqlRestApplication.java#L903

Reply:
The reason that that config is used in the TopicStreamWriter is explained here: https://github.com/confluentinc/ksql/blob/master/ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/KsqlRestConfig.java#L197-L199

This mechanism (writing an empty line to keep the connection alive) is not necessary for HTTP/2. Therefore, we don't need it in this new endpoint usage.

> Should we use a fixed value, something like 50 ms?

Let's choose something larger, in case the connection to Kafka is slow. We can emulate ScalablePushConsumer and choose 5 seconds: https://github.com/confluentinc/ksql/blob/91e24e9df047fa8b245562c927fa3a57e4f2b745/ksqldb-engine/src/main/java/io/confluent/ksql/execution/scalablepush/consumer/ScalablePushConsumer.java#L53

Reply:
Done!